CMakeLists.txt

```diff
@@ -10,564 +10,279 @@
 cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 project(valhalla LANGUAGES CXX C)
 
-list(INSERT CMAKE_MODULE_PATH 0 ${CMAKE_SOURCE_DIR}/cmake)
-
-set(boost_components
-  date_time
-  filesystem
-  iostreams
-  program_options
-  regex
-  system
-  thread)
-if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
-  list(APPEND boost_components zlib)
+include(GNUInstallDirs)
+
+set(VALHALLA_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+list(INSERT CMAKE_MODULE_PATH 0 ${VALHALLA_SOURCE_DIR}/cmake)
+
+option(ENABLE_CCACHE "Speed up incremental rebuilds via ccache" ON)
+option(ENABLE_DATA_TOOLS "Enable Valhalla data tools" ON)
+option(ENABLE_SERVICES "Enable Valhalla services" ON)
+option(ENABLE_PYTHON_BINDINGS "Enable Python bindings" ON)
+option(ENABLE_COVERAGE "Build with coverage instrumentalisation" OFF)
+option(ENABLE_SANITIZER "Use memory sanitizer for Debug build" OFF)
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+
+if(ENABLE_CCACHE AND (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID MATCHES "GNU"))
+  find_program(CCACHE_FOUND ccache)
+  if(CCACHE_FOUND)
+    message(STATUS "Using ccache to speed up incremental builds")
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
+    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
+    set(ENV{CCACHE_CPP2} "true")
+  endif()
 endif()
-find_package(Boost 1.51 REQUIRED COMPONENTS ${boost_components})
-include_directories(${Boost_INCLUDE_DIRS})
 
-find_package(CURL REQUIRED)
-include_directories(${CURL_INCLUDE_DIR})
+if(ENABLE_SANITIZER)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address")
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=address")
+  set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -fsanitize=address")
+endif()
 
-find_package(LZ4 REQUIRED)
-include_directories(${LZ4_INCLUDE_DIR})
 
-find_package(Lua REQUIRED)
-include_directories(${LUA_INCLUDE_DIR})
+## libvalhalla
+add_subdirectory(src)
 
-find_package(Protobuf REQUIRED)
-include_directories(${PROTOBUF_INCLUDE_DIRS})
 
-find_package(SQLite3 REQUIRED)
+## Executable targets
+find_package(Boost 1.51 REQUIRED COMPONENTS program_options)
 
-find_package(SpatiaLite)
+function(get_source_path PATH NAME)
```

Comment:
It might be a little bit easier to understand using code that is less magic here and just hard coding the path.